chassembly/src/main/java/org/quiltmc/chasm/lang/internal/parse/Parser.jj

```diff
@@ -1,517 +1,1414 @@
+{
+   default Node evaluate(Evaluator evaluator) {throw new UnsupportedOperationException("unimplemented");}
+   default void resolve(Resolver resolver) {throw new UnsupportedOperationException("unimplemented");}
+   default void render(Renderer renderer, StringBuilder builder, int currentIndentationMultiplier) {throw new UnsupportedOperationException("unimplemented");}
 }
 
-PARSER_BEGIN(Parser)
-
-import java.util.ArrayList;
-import java.util.LinkedHashMap;
-
-import org.quiltmc.chasm.lang.api.ast.BinaryNode;
-import org.quiltmc.chasm.lang.api.ast.BooleanNode;
-import org.quiltmc.chasm.lang.api.ast.CallNode;
-import org.quiltmc.chasm.lang.api.ast.FloatNode;
-import org.quiltmc.chasm.lang.api.ast.IndexNode;
-import org.quiltmc.chasm.lang.api.ast.IntegerNode;
-import org.quiltmc.chasm.lang.api.ast.LambdaNode;
-import org.quiltmc.chasm.lang.api.ast.ListNode;
-import org.quiltmc.chasm.lang.api.ast.NullNode;
-import org.quiltmc.chasm.lang.api.ast.StringNode;
-import org.quiltmc.chasm.lang.api.ast.MapNode;
-import org.quiltmc.chasm.lang.api.ast.MemberNode;
-import org.quiltmc.chasm.lang.api.ast.Node;
-import org.quiltmc.chasm.lang.api.ast.ReferenceNode;
-import org.quiltmc.chasm.lang.api.ast.TernaryNode;
-import org.quiltmc.chasm.lang.api.ast.UnaryNode;
-
-public class Parser {
+INJECT BaseNode :
+   import java.io.IOException;
+   import java.nio.file.Files;
+   import java.nio.file.Path;
+   import org.quiltmc.chasm.lang.internal.render.Renderer;
+{
+   /**
+     * Parse a given file into a {@link Node}.
+     *
+     * @param path Path to a valid source file containing a single node.
+     * @return The node parsed from the given source file.
+     * @throws IOException If there is an error reading the file.
+     * @throws ParseException If the source contains syntax errors.
+     */
+    public static Node parse(Path path) throws IOException {
+        Parser parser = new Parser(path);
+        return parser.file();
+    }
 
-}
+    /**
+     * Parse a given {@link String} into a {@link Node}.
+     *
+     * @param string A string containing a single node.
+     * @return The node parsed from the given source file.
+     * @throws ParseException If the source contains syntax errors.
+     */
+    public static Node parse(String string) {
+        Parser parser = new Parser(string);
+        return parser.file();
+    }
 
-PARSER_END(Parser)
+    public void write(Path path) throws IOException {
+        Renderer renderer = Renderer.builder().build();
+        StringBuilder sb = new StringBuilder();
+        render(renderer, sb, 1);
+        Files.write(path, sb.toString().getBytes()); // what about utf16 support?
+    }    
+}
 
 // Productions
-Node file():
+Node file :
 {
-    Node e;
+    Node n;
 }
-{
-    e = expression()
+    n = expression
     <EOF>
-    { return e; }
-}
+    { return n; }
+;
 
 // Expressions
 // Note: These are nested according to operator precedence
 
-Node literalExpression():
+INJECT IntegerNode : 
```

Comment:
I'm not a very big fan of all of the AST nodes being generated like this, and for something that is part of Chassembly's public API, it makes it needlessly more complicated to edit, what with having to go through a nearly 1.5k line long file.

Reply:
Well, frankly, your comment is based on a very basic misunderstanding. As best I can tell, you believe that there is an all-or-none choice being presented to use code injection to *generate* the nodes (ALL of them) OR to maintain the various XXXNode subclasses ALL by hand.

That is not the case. I did give you can example where all the XXXNode code was generated, but in fact, you can mix-and-match the two approaches. Here is a concrete example of what I'm talking about in internal development: https://github.com/revusky/javacc21/blob/master/src/javacc/JavaCC.javacc#L1497

      ExpansionSequence #com.javacc.core.ExpansionSequence :

The fact that the Node class is fully specified with package name is taken to mean that it is maintained by hand. And you can see that here: https://github.com/revusky/javacc21/blob/master/src/java/com/javacc/core/ExpansionSequence.java

So, any Node class that is getting fairly big, you can opt out of auto-generation and just have it be maintained by hand. There is one little caveat, which is that your hand-written Node class must extend BaseNode (directly or indirectly) or it could *implement* the Node interface, but that requires more boilerplate. Typically you just subclass BaseNode or something else that subclasses BaseNode. So, for example, the com.javacc.core.ExpansionSequence subclasses com.javacc.core.Expansion which is an abstract base class that itself subclasses BaseNode.

But look, here's the point. If you have a complex grammar, it is likely to generate a whole slew of Node subclasses. Just as an example, the Java parser included in JavaCC21 generates 158 node subclasses. Most of them are either trivial or we `INJECT` a method or two inside for convenience, like getter/setter type stuff. Well, just try it:
 
            git clone https://github.com/javacc21/javacc21.git
            cd javacc21
            ant jar
            cd examples/java
            ant test

At that point, you'd see that there are 158 different *generated* Node subclasses in `./org/parsers/java/ast`. If you look inside of them, you would see that most just have trivial boilerplate. And some have just a few methods *injected*. So, first of all, I would put it to you that the tool *generating* well over a hundred Node subclasses is a pretty good leg-up in terms of getting a project up and running.

But the point is that if you find yourself injecting too much code into a generated Node subclass, you can certainly opt to hand-maintain/edit that class. So you get the best of both worlds. You should actually, by the way, understand that what I am doing for you with this pull request is quite unnatural. I'm taking a situation where you've hand-written the nodes and refactoring it so that it's all generated. Normally, if you were using the tool as intended, you start with everything generated, and then if some particular nodes get unwieldy, you switch to hand-editing them. But as a practical matter, if you have a hundred or more Node types in your AST, most will be quite trivial, and there might only be a few that you choose to hand-edit.

As long as the Node class is quite small, it is much better for it to be generated. For example, take the example I gave of the `ExendsList` in the Java grammar. If you look at that point, here: https://github.com/javacc21/javacc21/blob/master/examples/java/Java.javacc#L229

You see that there is an `INJECT` right under it, just really it implements a bit of syntactic sugar, so you could write `extendsList.getTypes()`

Now, that is the only injected method in `ExtendsList`, so here is the `ExtendsList.java` that it generates for you:

    package org.parsers.java.ast;

    import org.parsers.java.*;
    import static org.parsers.java.JavaConstants.TokenType.*;
    import java.util.List;
    @SuppressWarnings("unused")
    public class ExtendsList extends BaseNode {
        public List<ObjectType> getTypes() {
            return childrenOfType(ObjectType.class);
        }
    }

I mean your idea (implicitly) is that it is somehow better to have to write over a hundred itty bitty classes like this and check them into git and so on. I mean, again, if a given class becomes so big, so much code is injected into it that you want to maintain it separately, you can do that, but there are huge advantages to having it all generated. And also it's part of an existing Node API that you can use, so you can do things like:

           List<ExtendsList> allExtendsLists =  compilationUnit.descendantsOfType(ExtendsList.class);

And you can get all that for free. If you want to have a Visitor that just walks over all the nodes and handles them in different "handlers", well, that's basically free too.

So you're leveraging the fact that there is already a Node API that the tool generates for you, and if you step back and think about it, is this not much more the way modern computing is supposed to work? Do you come into every project and write from scratch your own hash table and growable list? Surely not. You just use java.util.HashMap and java.util.ArrayList. So why do you think it makes sense to create your own Node API from scratch? And, again, even when you decide to hand-write a Node, you are still able to use the Node API. So, you would see things in the JavaCC21 codebase (that is written with JavaCC21) like:

            for (BNFProduction prod :  grammar.descendantsOfType(BNFProduction.class)) {
                     // Do something with prod
            }

This works because both com.javacc.Grammar and com.javacc.core.BNFProduction inherit from (the generated) BaseNode and we have all this API available to us for free. But look, if you like this a lot and decide to use JavaCC21 in your next project, you use the same generated Node API and you hit the ground running. You start off another project that is similar (but not quite the same) and you're going to write another Node API?

Well, I'm just making the case. You can reflect on this. I mean, look, you do have to consider the fact that I probably have thought about these issues more than you have!

Reply:
Oh, and another point is that the `INJECT` disposition does give you the ability to define your type hierarchy in the generated nodes, so consider something like: https://github.com/javacc21/javacc21/blob/master/examples/java/Java.javacc#L534-L578

That is also from the Java grammar. But, look, getting back to that `ExtendsList` example: https://github.com/javacc21/javacc21/blob/master/examples/java/Java.javacc#L229-L245

What the `INJECT` disposition gives us is the ability to define things very close to where they are used/relevant. I mean, somebody just reading the code I link above for ExtendsList, that person also sees immediately that the generated `ExtendsList` Node class also has this `getTypes()` convenience method that he can use. Because it's there right next to the `ExtendsList` BNF production! If the ExtendsList.java was maintained separately, then to know this, the person reading the grammar has to go off on some wild goose chase to find the information. I really think it is far better for developer productivity to have key information all in one place. But again, there is some trade-off in terms of generated vs hand-maintained, so if you decide a given Node class is better maintained by hand, you can do that too. But, again, in a system with well over a hundred node types, you really would only want to hand-edit a few of them probably. And that said, the trade-off could shift on that wonderful day in the future where there is good tooling for JavaCC21 and then you would have syntax highlighting and code completion for injected code as well... But even in the current situation, the ability to *inject* a handful of convenience methods in a generated Node subclass is extremely useful. That, actually, is not something I care to debate about really. I *know* this experientially!

But, hey, some people like doing things the hard way... LOL.

Reply:
> But, hey, some people like doing things the hard way... LOL.

Ok, look, I appreciate your help, but can you try to phrase your stuff a little less condescending? You make it sound like we're doing things completely wrong just because we don't fully understand a hardly documented project.